packages/multichain/src/scope/transform.ts

```diff
@@ -0,0 +1,145 @@
+
+/**
+ * Normalizes a ScopeString and ExternalScopeObject into a separate
+ * InternalScopeString and InternalScopeObject for each reference in the `references`
+ * value if defined and adds an empty `accounts` array if not defined.
+ *
+ * @param scopeString - The string representing the scope
+ * @param externalScopeObject - The object that defines the scope
+ * @returns a map of caipChainId to ScopeObjects
+ */
+export const normalizeScope = (
+  scopeString: string,
+  externalScopeObject: ExternalScopeObject,
+): InternalScopesObject => {
+  const { references, ...scopeObject } = externalScopeObject;
+  const { namespace, reference } = parseScopeString(scopeString);
+
+  const normalizedScopeObject: InternalScopeObject = {
+    accounts: [],
+    ...scopeObject,
+  };
+
+  const shouldFlatten =
+    namespace &&
+    !reference &&
+    references !== undefined &&
+    references.length > 0;
+
+  if (shouldFlatten) {
+    return Object.fromEntries(
+      references.map((ref: CaipReference) => [
+        `${namespace}:${ref}`,
+        cloneDeep(normalizedScopeObject),
+      ]),
+    );
+  }
+  return { [scopeString]: normalizedScopeObject };
+};
+
+export const mergeScopeObject = (
+  scopeObjectA: InternalScopeObject,
+  scopeObjectB: InternalScopeObject,
+) => {
+  const mergedScopeObject: InternalScopeObject = {
+    methods: getUniqueArrayItems([
+      ...scopeObjectA.methods,
+      ...scopeObjectB.methods,
+    ]),
+    notifications: getUniqueArrayItems([
+      ...scopeObjectA.notifications,
+      ...scopeObjectB.notifications,
+    ]),
+    accounts: getUniqueArrayItems([
+      ...scopeObjectA.accounts,
+      ...scopeObjectB.accounts,
+    ]),
+  };
+
+  if (scopeObjectA.rpcDocuments || scopeObjectB.rpcDocuments) {
+    mergedScopeObject.rpcDocuments = getUniqueArrayItems([
+      ...(scopeObjectA.rpcDocuments ?? []),
+      ...(scopeObjectB.rpcDocuments ?? []),
+    ]);
+  }
+
+  if (scopeObjectA.rpcEndpoints || scopeObjectB.rpcEndpoints) {
+    mergedScopeObject.rpcEndpoints = getUniqueArrayItems([
+      ...(scopeObjectA.rpcEndpoints ?? []),
+      ...(scopeObjectB.rpcEndpoints ?? []),
+    ]);
+  }
+
+  return mergedScopeObject;
+};
+
+/**
+ * Merges two InternalScopeObjects
+ * @param scopeA - The first scope object to merge.
+ * @param scopeB - The second scope object to merge.
+ * @returns The merged scope object.
+ */
+export const mergeScopes = (
+  scopeA: InternalScopesObject,
+  scopeB: InternalScopesObject,
+): InternalScopesObject => {
+  const scope: InternalScopesObject = {};
+
+  Object.entries(scopeA).forEach(([_scopeString, scopeObjectA]) => {
+    const scopeString = _scopeString as InternalScopeString;
+    const scopeObjectB = scopeB[scopeString];
+
+    scope[scopeString] = scopeObjectB
+      ? mergeScopeObject(scopeObjectA, scopeObjectB)
+      : scopeObjectA;
+  });
+
+  Object.entries(scopeB).forEach(([_scopeString, scopeObjectB]) => {
+    const scopeString = _scopeString as InternalScopeString;
```

Comment:
Nit: This cast is unfortunate but I don't see a great alternative without adding a redundant runtime check, which would be silly because TypeScript ought to know this, but it's lost during some type flattening happening in `Object.entries`.

Instead of casting like this, I'd recommend using `keyof typeof [object]` so we at least get some warning when the type of the parameter changes. Otherwise this might just silently keep applying the wrong type. I'd also suggest a comment explaining why the cast is needed


```suggestion
    // Cast needed because index type is returned as `string` by `Object.entries`
    const scopeString = _scopeString as keyof typeof scopeA;
    const scopeObjectB = scopeB[scopeString];

    scope[scopeString] = scopeObjectB
      ? mergeScopeObject(scopeObjectA, scopeObjectB)
      : scopeObjectA;
  });

  Object.entries(scopeB).forEach(([_scopeString, scopeObjectB]) => {
     // Cast needed because index type is returned as `string` by `Object.entries`
    const scopeString = _scopeString as keyof typeof scopeB;
```